cpp/include/cugraph/edge_properties.hpp

```diff
@@ -0,0 +1,210 @@
+    auto& prop = std::get<raft::device_span<int32_t>>(property);
+    return func(prop);
+  } else if (std::holds_alternative<raft::device_span<int64_t>>(property)) {
+    auto& prop = std::get<raft::device_span<int64_t>>(property);
+    return func(prop);
+  } else if (std::holds_alternative<raft::device_span<size_t>>(property)) {
+    auto& prop = std::get<raft::device_span<size_t>>(property);
+    return func(prop);
+  }
+
+  CUGRAPH_FAIL("unsupported variant type -- shouldn't happen");
+}
+
+template <typename func_t>
+auto variant_type_dispatch(const_numeric_device_span_t& property, func_t func)
+{
+  if (std::holds_alternative<raft::device_span<float const>>(property)) {
+    auto& prop = std::get<raft::device_span<float const>>(property);
+    return func(prop);
+  } else if (std::holds_alternative<raft::device_span<double const>>(property)) {
+    auto& prop = std::get<raft::device_span<double const>>(property);
+    return func(prop);
+  } else if (std::holds_alternative<raft::device_span<int32_t const>>(property)) {
+    auto& prop = std::get<raft::device_span<int32_t const>>(property);
+    return func(prop);
+  } else if (std::holds_alternative<raft::device_span<int64_t const>>(property)) {
+    auto& prop = std::get<raft::device_span<int64_t const>>(property);
+    return func(prop);
+  } else if (std::holds_alternative<raft::device_span<size_t const>>(property)) {
+    auto& prop = std::get<raft::device_span<size_t const>>(property);
+    return func(prop);
+  }
+
+  CUGRAPH_FAIL("unsupported variant type -- shouldn't happen");
+}
+
+template <typename edge_t, typename func_t>
+auto variant_type_dispatch(edge_property_views_t<edge_t>& property, func_t func)
+{
+  if (std::holds_alternative<cugraph::edge_property_view_t<edge_t, float const*>>(property)) {
+    auto& prop = std::get<cugraph::edge_property_view_t<edge_t, float const*>>(property);
+    return func(prop);
+  } else if (std::holds_alternative<cugraph::edge_property_view_t<edge_t, double const*>>(
+               property)) {
+    auto& prop = std::get<cugraph::edge_property_view_t<edge_t, double const*>>(property);
+    return func(prop);
+  } else if (std::holds_alternative<cugraph::edge_property_view_t<edge_t, int32_t const*>>(
+               property)) {
+    auto& prop = std::get<cugraph::edge_property_view_t<edge_t, int32_t const*>>(property);
+    return func(prop);
+  } else if (std::holds_alternative<cugraph::edge_property_view_t<edge_t, int64_t const*>>(
+               property)) {
+    auto& prop = std::get<cugraph::edge_property_view_t<edge_t, int64_t const*>>(property);
+    return func(prop);
+  } else if (std::holds_alternative<cugraph::edge_property_view_t<edge_t, size_t const*>>(
+               property)) {
+    auto& prop = std::get<cugraph::edge_property_view_t<edge_t, size_t const*>>(property);
+    return func(prop);
+  }
+
+  CUGRAPH_FAIL("unsupported variant type -- shouldn't happen");
+}
+
+struct variant_size {
```

Comment:
I think this name is confusing, I first thought this is about the number of data types a variant object can hold. I will revisit this once I fully reviewed this PR.

Reply:
Renamed in next push.  I used `sizeof_arithmetic_element`

Reply:
Sorry for nitpicking but sizeof_arithment_element sounds like sizeof(T) for rmm::device_uvector<T>...

Reply:
Oh, sorry, this is actually sizeof(T), great name!!!